### PR TITLE
[cudagraph] Capture pipeline forward-backward steps

### PR DESCRIPTION
**Pull request by xmfan**
Stacked PRs:
 * #4485
 * __->__#4435
 * #4433
 * #4432
 * #4430

--- --- ---

### [cudagraph] Capture pipeline forward-backward steps

Note: this doesn't add looped schedule support yet, that will need the directed p2p communicators (https://github.com/pytorch/pytorch/pull/195760), a reference of adding support is in https://github.com/pytorch/torchtitan/pull/4485.

## Why

TorchTitan captures the non-pipeline forward-backward function, but pipeline training still executes every schedule eagerly. A pipeline schedule with one stage per rank is a suitable graph boundary once input preprocessing remains outside capture and communication order, process groups, tensor structure, and tensor metadata remain static.

This change adds CUDA graph capture for single-stage-per-rank pipeline schedules. Looped schedules remain unsupported here because they require separate directed P2P communicators.

## Capture boundary

```text
Eager code
    |
    +--> load microbatches
    +--> move tensors to the device
    +--> preprocess model inputs
    |
    v
CUDA graph boundary
    |
    +--> pipeline schedule
    +--> forward computation
    +--> pipeline P2P communication
    +--> loss computation
    +--> backward computation
    +--> detached microbatch-loss reduction
    |
    v
Eager code
    |
    +--> gradient clipping
    +--> optimizer and scheduler steps
    +--> metrics reduction and logging
```

## Change

Split pipeline preprocessing from `_pp_forward_backward_body()` and select either the pipeline or non-pipeline body through the existing `fwd_bwd_fn` before applying the CUDA graph wrapper. Use a persistent loss sentinel on ranks that do not own the last stage.

Reject looped pipeline schedules until their directed P2P communicator dependency is available, and reject pipeline CUDA graphs with validation because validation reinitializes the shared schedule and invalidates the captured training graph.

## Testing

The unit tests cover configuration validation, structured pipeline inputs, persistent loss output, and wrapper selection.

- `python -m pytest -q tests/unit_tests/cpu/test_cudagraph.py tests/unit_tests/cpu/test_trainer.py tests/unit_tests/cpu/test_config_manager.py tests/unit_tests/cpu/test_integration_test_definitions.py`
- `PYTHONPATH=. python tests/integration_tests/run_tests.py <output> --test_suite features --test_name pp_1f1b_cudagraph --execution_mode real_pg --ngpu 2 --no-parallel` (10/10 steps)

### CODE REVIEW

File: tests/integration_tests/features.py [hunk 1]
```diff
@@ -110,6 +110,14 @@ def build_features_test_list() -> list[OverrideDefinitions]:
             ngpu=2,
             use_real_pg=True,
         ),
+        OverrideDefinitions(
```
Comment (tianyu-l) [inline tests/integration_tests/features.py:113]: instead of adding new one, can we just remove the disable-cuda-graph flag in these two existing tests https://github.com/pytorch/torchtitan/blob/main/tests/integration_tests/features.py#L106-L121

File: torchtitan/trainer.py
Comment (tianyu-l) [inline torchtitan/trainer.py:825]: This function becomes pretty short now. To reduce surface, can we inline it back to `forward_backward_step`?

File: torchtitan/trainer.py [hunk 2]
```diff
@@ -289,7 +306,8 @@ def maybe_log(self) -> None:
     device: torch.device
     gc_handler: utils.GarbageCollection
     train_context: dist_utils.SpmdContext
-    fwd_bwd_fn: ForwardBackwardFn
+    fwd_bwd_fn: Callable[..., torch.Tensor]
+    _pp_loss_sentinel: torch.Tensor
```
Comment (tianyu-l) [inline torchtitan/trainer.py:310]: to make it more readable (at the cost of being verbose)
```suggestion
    _pp_loss_sentinel_on_non_last_stage: torch.Tensor
```
or whatever you'd propose alternatively.

We probably should broadcast loss to every PP rank though. Not a blocker for this PR

**Review by tianyu-l — APPROVED** [review]
